### PR TITLE
storage: pass limit param as hint in querier

### PR DESCRIPTION
**Pull request by harry671003**
### Related Issues
- Issue: #12795
- Prior work:  #13396

### Description
In #13396 a new limit param for the series, label names and label values APIs was introduced. This limit is currently only applied in the api.go. The storage implementations that do not support streaming such as Thanos, Cortex etc, will not be able to receive the limit param and would have to return all the results.

This change would pass the limit param as hints to storage so that the implementations can use that to return only the required number of results.

### CODE REVIEW

File: storage/merge_test.go
Comment (machine424) [inline storage/merge_test.go:1563]: The only changes we should make to those tests is set the new arg `hints=nil` and everything should keep working as before.
To test the limit hint, I suggest you add new tests/cases.

File: storage/interface.go
Comment (machine424) [inline storage/interface.go:193]: Maybe we should explicitly mention that `0` means no limit (for minimal changes)

File: web/api/v1/api.go
Comment (machine424) [inline web/api/v1/api.go:699]: Note that the warnings logic, the `if len(names) >= limit {` block in 709 is broken, https://github.com/prometheus/prometheus/pull/14116 is to fix that.

If that PR is merged before, those new blocks will become unreachable (if the implementations respect the limit, which is expected).

We'll need to think about how to handle that, maybe ask for `limit+1`...
(same applies to labelsValues and series I think).

**Review by machine424 — CHANGES_REQUESTED** [review]
Thanks for this.

AFAIU, the limit hint isn't taken into account by the `mergeGenericQuerier` implementation e.g. 
For the `Select` it's not that important as it streams, but we'll need that for the labels fcts.
We could create an issue for that and address it in another PR.

**Comment by machine424** [comment]
After this is in place, we could also make use of that limit hint for `storage.remote.read-sample-limit` https://github.com/prometheus/prometheus/pull/4532/files

File: web/api/v1/api.go
Comment (harry671003) [reply web/api/v1/api.go:701]: Good catch. I think requesting for limit+1 makes sense to keep the changes minimal.

File: storage/merge_test.go
Comment (harry671003) [reply storage/merge_test.go]: Yes, that makes sense. I would add more tests when I implement the limiting logic inside the `mergeGenericQuerier`. I will create a follow up PR for that soon after this is merged.

**Comment by harry671003** [comment]
Hey @bboreham, could you please take a look at this PR?

**Comment by machine424** [comment]
> > After this is in place, we could also make use of that limit hint for `storage.remote.read-sample-limit` https://github.com/prometheus/prometheus/pull/4532/files
> 
> @machine424 - Sorry, I didn't understand how the sample limit is relevant here. Could you elaborate more?

We can, later, set the limit hint here https://github.com/prometheus/prometheus/blob/1081e336a0c4004b6f817c70f0ff4f3810db80fa/storage/remote/read_handler.go#L147-L161 in accordance with `sampleLimit` if it makes sense.

---

You can rebase as https://github.com/prometheus/prometheus/pull/14116 was merged.

File: web/api/v1/api.go
Comment (machine424) [inline web/api/v1/api.go:662]: Actually if no limit is set, `limit` will be `math.MaxInt` and this will overflow.
That case should be handled separately.  
Maybe we can add a `limitParamToLimitHint` util or sth.

Comment (machine424) [reply web/api/v1/api.go:662]: If we can align the meaning of "no limit" between them, it'd be even better, we can make `parseLimitParam` consider 0 as the no limit value (as we do for the hint) + add some doc comments to state that, we'll need to turn the `if len(...) > limit` checks into  `if limit >0 && len(...) > limit` (we can have an inline util for that), but at least in this way there is no confusion and no conversion to be done and the "no limit" can return more than `9*10^18` :)

File: storage/interface.go
Comment (bboreham) [inline storage/interface.go:194]: Please format this comment consistently with other members of the struct.

File: storage/interface.go
Comment (bboreham) [inline storage/interface.go:164]: Changing this interface seems like quite a big step; it forces a change on all downstream implementations.

**Review by bboreham — COMMENTED** [review]
If I read this correctly you didn't implement early exit when the limit was hit; just added the interface so that someone else could do that.
This seems wrong to me; if we're going to add this complexity we should make Prometheus better too.

File: storage/interface.go
Comment (harry671003) [reply storage/interface.go:164]: Thanks for reviewing. I was also hesitant to change the interface initially. But only other way I could think of passing the limit to storage was through the `context`.

I also noticed that the interface was changed a few times in the past to add the matchers to the API.
- eg: https://github.com/prometheus/prometheus/pull/8400

Is this a blocker? Does it need more discussions with the maintainers?

**Comment by harry671003** [comment]
> If I read this correctly you didn't implement early exit when the limit was hit; just added the interface so that someone else could do that. This seems wrong to me; if we're going to add this complexity we should make Prometheus better too.

I was afraid that including this change in this PR would make it big. I will create a PR soon after this to implement those changes.

**Comment by machine424** [comment]
> If I read this correctly you didn't implement early exit when the limit was hit; just added the interface so that someone else could do that. This seems wrong to me; if we're going to add this complexity we should make Prometheus better too.

Yes, we discussed that for `mergeGenericQuerier` here https://github.com/prometheus/prometheus/pull/14109#pullrequestreview-2064329232, the labels related ones would be more needed as that implementation's `Select` is an iteration. Did we miss any other implementation?
(If we agree on that,  let's mention that in the PR description, open en issue and link it there as well)

File: web/api/v1/api_test.go
Comment (machine424) [inline web/api/v1/api_test.go:781]: Is it guaranteed that they'll always be the same?

File: web/api/v1/api_test.go
Comment (machine424) [inline web/api/v1/api_test.go:749]: I'd add a `limit!=""` check before this so we don't change the previous behavior. (before we weren't setting the limit param, now we're setting it to the empty string, they should be handled the same way, but it's better no to change existing tests)

**Review by machine424 — COMMENTED** [review]
lgtm so far, just some nits.

File: web/api/v1/api.go
Comment (harry671003) [reply web/api/v1/api.go:711]: Sorry, I'm not sure I understand. 

In the approach I went with, the default value of limit would be 0. If the code doesn't check for `limit > 0`, then if the default case, won't no results be returned?

```
if len(names) > limit {
  vals = vals[:limit]
  warnings = warnings.Add(errors.New("results truncated due to limit"))
}
```

File: web/api/v1/api_test.go
Comment (harry671003) [reply web/api/v1/api_test.go:781]: IIUC, the response from the querier is sorted. So the first two entries should always be the same? Please let me know if that's not the case.

I also ran the test 100K times just to check if it always passes.

Comment (bboreham) [reply web/api/v1/api_test.go:783]: I wasn't sure, so looked it up: https://github.com/prometheus/prometheus/blob/f5eff89811cb3249f5fb3345e3a2350936b0362c/tsdb/block.go#L98
https://github.com/prometheus/prometheus/blob/f5eff89811cb3249f5fb3345e3a2350936b0362c/storage/interface.go#L166

File: web/api/v1/api.go
Comment (machine424) [reply web/api/v1/api.go:711]: Ah, ok so you went for the 2nd approach: align API and storage no-limit meaning.

`getStorageLimitFromLimitParam` has misled me.

In this case, it should also handle the case where API `limit=0`, it should ask the storage for `1`, how about

```
// toStorageLimit increases the API limit, as returned by parseLimitParam, by 1. This allows for emitting warnings when the results are truncated.
func toStorageLimit(limit int) int {
        // 0 means no limit and avoid int overflow
	if limit > 0 && limit < math.MaxInt {
		return limit + 1
	}
	return limit
}
```
?

It'd be great to add a `limit=0` case to `testEndpoints` (it'll be relevant once the storage limit is implemented).

(I know that handling corner cases is not the most pleasant thing to do, but we’re very close to the end ;))

File: web/api/v1/api.go
Comment (machine424) [inline web/api/v1/api.go:1910]:
```suggestion
// parseLimitParam returning 0 means no limit is to be applied.
func parseLimitParam(limitStr string) (limit int, err error) {
```

File: web/api/v1/api_test.go
Comment (machine424) [reply web/api/v1/api_test.go:783]: Yes, I'm a lazybones :) Thanks to you two.

File: web/api/v1/api.go
Comment (harry671003) [reply web/api/v1/api.go:711]: Oh, I missed that edgecase :/ Thanks for catching this.

**Review by machine424 — DISMISSED** [review]
thanks again!
looking forward to the implementation PR.
Let's see if it's ok for Bryan as well.

**Comment by harry671003** [comment]
I took a stab at implementing the limits inside merge.go
I included those changes in the same PR as changes seemed minimal.
Could you take another look @machine424 and @bboreham ?

File: storage/merge.go
Comment (harry671003) [reply storage/merge.go:173]: The hints needed to be passed inside `lvals()` so that it can be passed to the `LabelValues()` invocation on each querier.

**Comment by machine424** [comment]
> I took a stab at implementing the limits inside merge.go I included those changes in the same PR as changes seemed minimal. Could you take another look @machine424 and @bboreham ?

I'd proceed as we agreed: merge the PR with the interface changes only (I already approved those) if no one has an objection, and then add the implementation in another PR. This will also keep reviewing easy.

**Comment by harry671003** [comment]
> I'd proceed as we agreed: merge the PR with the interface changes only (I already approved those) if no one has an objection, and then add the implementation in another PR.

I reverted back to the original. I can create a separate PR for the implementation. :)

File: web/api/v1/api.go
Comment (bboreham) [inline web/api/v1/api.go:1929]: This feels a bit weird to me.
After some thought, I realized it allows the feature to work on storages which implement the limit and those which do not.  Is that why you did it?
Please write down why this behaviour exists.

Also I don't like the name; it's got nothing to do with "storage".  Maybe "toHintLimit" ?

Comment (harry671003) [reply web/api/v1/api.go:1929]: The warning `results truncated due to limit` is emitted in `api.go` only when the number of results returned from `storage.Querier` is greater than the limit.
To ensure that this logic works correctly, limit + 1 should be passed to the `storage.Querier`.

The intend with storage limit is to mean the limit that is passed to `storage.Querier`. I don't mind renaming it to `toHintLimit`.

**Review by machine424 — DISMISSED** [review]
lgtm, Let's see if it's ok for Bryan.

File: web/api/v1/api.go
Comment (bboreham) [inline web/api/v1/api.go:1928]: ... when used with storages which do not implement the limit ?

Comment (machine424) [reply web/api/v1/api.go]: This will also be used for storages that implement the limit.
Asking the storage for `user limit+1` let us decide if  with `user limit`, the user sees truncated results.

Comment (machine424) [reply web/api/v1/api.go]: To continue being able to emit these warnings:
https://github.com/prometheus/prometheus/blob/f24ce00320f8605add7c7c5e69f722193c066672/web/api/v1/api.go#L892-L894

Maybe we should make `This allows for emitting warnings when the results are truncated.` clearer.

Comment (machine424) [reply web/api/v1/api.go]: (That being said, we can think about whether we can make the storage itself send such warnings, but maybe in another PR)

Comment (harry671003) [reply web/api/v1/api.go:1928]: > ... when used with storages which do not implement the limit ?

After some thought, adding this to the comment seems incorrect. The ONLY reason for this logic is to make the warnings work correctly. If a storage doesn't implement limit, then the warning will work correctly. Only when a storage implements the limit and returns the limit-ed results, the logic for emitting warnings will break.

**Comment by harry671003** [comment]
Could this PR be merged if there are no objections?

**Review by bboreham — APPROVED** [review]
Thanks!

**Pull request merged**
